### PR TITLE
implement atomic_flag_test_and_set(_explicit) and atomic_flag_clear(_explicit)

### PR DESCRIPTION
**Pull request by FMarno**
This contribution is being made by Codeplay on behalf of Samsung.

There are a few interesting design choices that I think could be disagreed with so I'll try to make my choices clear. The thing that is a little controversial is allowing the use `atomic_flag_test_and_set` and `atomic_flag_clear` since their default order is SequentiallyConsistent and that is treated as AcquireRelease in Vulkan shaders. Currently I add a warning to the compiler when those functions are used, but we might want to increase that to an error.

> All <id> used for Scope <id> and Memory Semantics <id> must be of an OpConstant.

https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#ShaderValidation
due to this we can expect the arguments to be known at compile time, this always us to do some static analysis and reject cases that are illegal. Due to this I force the compiler to panic when illegal scopes or semantics are used. The exception is memory_order_seq_cst.

>  (when using atomic_flag_clear) The order argument shall not be memory_order_acquire nor memory_order_acq_rel

https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_C.html#atomic_flag_clear

The compiler will crash when `memory_order_acquire` or `memory_order_acq_rel` are used with `atomic_flag_clear_explicit`. I've also used `memory_order_release` as the default for `atomic_flag_clear`, but it might be better to just reject the use of that function.

> Whilst the CrossDevice scope is defined in SPIR-V, it is disallowed in Vulkan

https://registry.khronos.org/vulkan/specs/1.3-khr-extensions/html/vkspec.html#shaders-scope-cross-device
The compiler will crash when `memory_scope_all_svm_devices`/`memory_scope_all_devices` is used with `atomic_flag_test_and_set_explicit` or `atomic_flag_clear_explicit`.

>  memory_scope_work_item can only be used with atomic_work_item_fence with flags set to CLK_IMAGE_MEM_FENCE. Requires support for OpenCL C 2.0 or newer.

https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_C.html#memory-scope
The compiler will crash when `memory_scope_work_item` is used with `atomic_flag_test_and_set_explicit` or `atomic_flag_clear_explicit`.

>  The functions not ending in explicit have the same semantics as the corresponding explicit function with memory_order_seq_cst for the memory_order argument.
> The functions that do not have memory_scope argument have the same semantics as the corresponding functions with the memory_scope argument set to memory_scope_device.

https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_C.html#atomic-functions

> The SPIR-V SubgroupMemory, CrossWorkgroupMemory, and AtomicCounterMemory memory semantics are ignored. Sequentially consistent atomics and barriers are not supported and SequentiallyConsistent is treated as AcquireRelease. SequentiallyConsistent should not be used.

https://registry.khronos.org/vulkan/specs/1.3-khr-extensions/html/vkspec.html#shaders-execution-memory-ordering

`memory_order_seq_cst` is treated as `memory_order_acq_rel` (or `memory_order_release` for `atomic_flag_clear`) and a warning is added where it's used.
`memory_scope_sub_group` is strengthened to `memory_scope_work_group` when used with `atomic_flag_test_and_set_explicit` or `atomic_flag_clear_explicit` and a warning is added were it's used.

### CODE REVIEW

**Comment by alan-baker** [comment]
The macos failures require a rebase to pickup the right script changes.

**Review by alan-baker — COMMENTED** [review]
Just some nits. I really like the diagnostics, thanks.

File: lib/ReplaceOpenCLBuiltinPass.cpp [hunk 1]
```diff
@@ -163,6 +167,26 @@ std::set<Builtins::BuiltinType> ReplaceOpenCLBuiltinPass::ReplaceableBuiltins =
...
+
+// https://registry.khronos.org/vulkan/specs/1.3-khr-extensions/html/vkspec.html#shaders-execution-memory-ordering
+// The SPIR-V SubgroupMemory, CrossWorkgroupMemory, and AtomicCounterMemory
+// memory semantics are ignored.
+enum class AtomicMemoryScope : uint32_t {
+  kMemoryScopeWorkItem = 0,
+  kMemoryScopeWorkGroup = 1,
+  kMemoryScopeDevice = 2,
+  kMemoryScopeAllSVMDevices = 3, // not supported
+  kMemoryScopeSubGroup = 4       // not supported
```
[the remark this answers is not in the document]
Comment (FMarno) [reply lib/ReplaceOpenCLBuiltinPass.cpp]: yes, and the comment above the enum. fixed!

File: lib/ReplaceOpenCLBuiltinPass.cpp [hunk 2]
```diff
@@ -3820,6 +3843,60 @@ bool ReplaceOpenCLBuiltinPass::replaceAtomicCompareExchange(Function &F) {
...
+    bool is_global = flag_pointer->getType()->getPointerAddressSpace() ==
+                     clspv::AddressSpace::Global;
+    auto semantics = MemoryOrderSemantics(
+        order_arg, is_global, Call, spv::MemorySemanticsAcquireReleaseMask);
+
+    auto scope = MemoryScope(scope_arg, is_global, Call);
+    IRBuilder<> builder(Call);
+    auto set_value = builder.getInt32(1);
+    auto previous_value = InsertSPIRVOp(
+        Call, spv::OpAtomicExchange, {Attribute::Convergent},
```
[the remark this answers is not in the document]
Comment (FMarno) [reply lib/ReplaceOpenCLBuiltinPass.cpp]: I also don't think it's necessary. I've removed it from the ones I've added but not touched the others.

**Review by alan-baker — APPROVED** [review]

**Pull request merged**